test/e2e/pv_reclaimpolicy.go

```diff
@@ -0,0 +1,158 @@
+	)
+
+	framework.KubeDescribe("persistentvolumereclaim:vsphere", func() {
+
+		BeforeEach(func() {
+			framework.SkipUnlessProviderIs("vsphere")
+			c = f.ClientSet
+			ns = f.Namespace.Name
+			framework.ExpectNoError(framework.WaitForAllNodesSchedulable(c, framework.TestContext.NodeSchedulableTimeout))
+		})
+
+		AfterEach(func() {
+			vsp, err := vsphere.GetVSphere()
+			Expect(err).NotTo(HaveOccurred())
+			testCleanupVSpherePersistentVolumeReclaim(vsp, c, ns, volumePath, pv, pvc)
+		})
+
+		/*
+			This test verifies persistent volume should be deleted when reclaimPolicy on the PV is set to delete and
+			associated claim is deleted
+
+			Test Steps:
+			1. Create vmdk
+			2. Create PV Spec with volume path set to VMDK file created in Step-1, and PersistentVolumeReclaimPolicy is set to Delete
+			3. Create PVC with the storage request set to PV's storage capacity.
+			4. Wait for PV and PVC to bound.
+			5. Delete PVC
+			6. Verify PV is deleted automatically.
+		*/
+		It("should delete persistent volume when reclaimPolicy set to delete and associated claim is deleted", func() {
+			framework.SkipUnlessProviderIs("vsphere")
+			vsp, err := vsphere.GetVSphere()
+			Expect(err).NotTo(HaveOccurred())
+
+			volumePath, pv, pvc, err = testSetupVSpherePersistentVolumeReclaim(vsp, c, ns, v1.PersistentVolumeReclaimDelete)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("verify pv is deleted")
+			err = framework.WaitForPersistentVolumeDeleted(c, pv.Name, 3*time.Second, 300*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		/*
+			This test Verify persistent volume should be retained when reclaimPolicy on the PV is set to retain
+			and associated claim is deleted
+
+			Test Steps:
+			1. Create vmdk
+			2. Create PV Spec with volume path set to VMDK file created in Step-1, and PersistentVolumeReclaimPolicy is set to Retain
+			3. Create PVC with the storage request set to PV's storage capacity.
+			4. Wait for PV and PVC to bound.
+			5. Delete PVC
+			6. Verify PV is retained.
+			7. Delete PV.
+		*/
+
+		It("should retain persistent volume when reclaimPolicy set to retain when associated claim is deleted", func() {
+			framework.SkipUnlessProviderIs("vsphere")
+			vsp, err := vsphere.GetVSphere()
+			Expect(err).NotTo(HaveOccurred())
+
+			volumePath, pv, pvc, err = testSetupVSpherePersistentVolumeReclaim(vsp, c, ns, v1.PersistentVolumeReclaimRetain)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("verify pv is retained")
+			framework.Logf("Waiting for PV %v to become Released", pv.Name)
+			err = framework.WaitForPersistentVolumePhase(v1.VolumeReleased, c, pv.Name, 3*time.Second, 300*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("delete pv")
+			deletePersistentVolume(c, pv.Name)
+			err = framework.WaitForPersistentVolumeDeleted(c, pv.Name, 3*time.Second, 300*time.Second)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("delete volume")
+			err = vsp.DeleteVolume(volumePath)
+			Expect(err).NotTo(HaveOccurred())
+		})
```

Comment:
The AfterEach deletes the vsp, pvc and pv. In this `It` the pvc, pv, and vsp are deleted before the AfterEach is invoked. The delete func invoked by the 'AfterEach' are resilient to the objects already being deleted, but the code here seems either redundant or misplaced. I'd either do all of the deleting in the `AfterEach` or in the `It` but not both.